File: engine/compose/src/commonMain/kotlin/dev/getmilano/expression/ExprAst.kt
package dev.getmilano

/** The expression AST, per the expression language spec's EBNF. */
internal sealed class Expr {
    data object NullLiteral : Expr()

    data class BoolLiteral(
        val value: Boolean,
    ) : Expr()

    data class IntLiteral(
        val value: Long,
    ) : Expr()

    data class DoubleLiteral(
        val value: Double,
    ) : Expr()

    data class StringLiteral(
        val value: String,
    ) : Expr()

    /** A reserved root: state, context, or event. */
    data class Root(
        val name: String,
    ) : Expr()

    data class Member(
        val base: Expr,
        val field: String,
    ) : Expr()

    data class Call(
        val name: String,
        val arguments: List<Expr>,
    ) : Expr()

    data class Unary(
        val op: UnaryOp,
        val operand: Expr,
    ) : Expr()

    data class Binary(
        val op: BinaryOp,
        val left: Expr,
        val right: Expr,
    ) : Expr()
}

internal enum class UnaryOp { NOT, NEGATE }

internal enum class BinaryOp {
    MULTIPLY,
    DIVIDE,
    MODULO,
    ADD,
    SUBTRACT,
    LESS,
    LESS_EQUAL,
    GREATER,
    GREATER_EQUAL,
    EQUAL,
    NOT_EQUAL,
    AND,
    OR,
    COALESCE,
}

/** A static expression error, mapped to SchemaViolation at the gate. */
internal class ExprException(
    val detail: String,
) : Exception(detail)

File: engine/swiftui/Sources/MilanoSDK/Expression/ExprAST.swift
import Foundation

/// The expression AST, per the expression language spec's EBNF.
indirect enum Expr: Equatable, Sendable {
    case nullLiteral
    case boolLiteral(Bool)
    case intLiteral(Int64)
    case doubleLiteral(Double)
    case stringLiteral(String)
    /// A reserved root: `state`, `context`, or `event`.
    case root(String)
    case member(Expr, String)
    case call(String, [Expr])
    case unary(UnaryOp, Expr)
    case binary(BinaryOp, Expr, Expr)
}

enum UnaryOp: Equatable, Sendable { case not, negate }

enum BinaryOp: Equatable, Sendable {
    case multiply, divide, modulo
    case add, subtract
    case less, lessEqual, greater, greaterEqual
    case equal, notEqual
    case and, or
    case coalesce
}

/// A static expression error, mapped to SchemaViolation at the gate.
struct ExprError: Error {
    let detail: String
}
